File: exercises/dotnet/ShippingService/Domain/Models/Shipment.cs
namespace ShippingService.Domain.Models;

/// <summary>
/// The Shipment aggregate root. In the workshop's model, shipping always
/// succeeds -- there is no failed or rejected state. No persistence;
/// created in-memory and returned.
/// </summary>
public record Shipment(
    ShipmentId Id,
    string OrderId,
    string ShippingClass,
    int EstimatedDays,
    DateTime ScheduledAt)
{
    /// <summary>Estimated delivery days by shipping class.</summary>
    private static readonly Dictionary<string, int> EstimatedDaysMap = new(StringComparer.OrdinalIgnoreCase)
    {
        ["overnight"] = 1,
        ["express"] = 2,
        ["priority"] = 3,
        ["standard"] = 5,
    };

    private const int DefaultEstimatedDays = 5;

    /// <summary>
    /// Factory: schedule a new shipment. Estimated delivery days are derived
    /// from the shipping class.
    /// </summary>
    public static Shipment Schedule(string orderId, string shippingClass)
    {
        var days = EstimatedDaysMap.GetValueOrDefault(shippingClass, DefaultEstimatedDays);
        return new Shipment(
            ShipmentId.Generate(),
            orderId,
            shippingClass,
            days,
            DateTime.UtcNow);
    }
}

File: exercises/quarkus/shipping-service/src/main/java/com/example/shipping/domain/model/Shipment.java
package com.example.shipping.domain.model;

import java.time.Instant;
import java.util.Objects;

public record Shipment(
        ShipmentId id,
        String orderId,
        String shippingClass,
        int estimatedDays,
        Instant scheduledAt
) {
    public Shipment {
        Objects.requireNonNull(id, "id");
        Objects.requireNonNull(orderId, "orderId");
        Objects.requireNonNull(shippingClass, "shippingClass");
        Objects.requireNonNull(scheduledAt, "scheduledAt");
    }

    public static Shipment schedule(String orderId, String shippingClass) {
        int days = switch (shippingClass.toLowerCase()) {
            case "overnight" -> 1;
            case "express" -> 2;
            case "priority" -> 3;
            default -> 5;
        };
        return new Shipment(ShipmentId.generate(), orderId, shippingClass, days, Instant.now());
    }
}
